services/sharedchannel/sync_send_remote.go

```diff
@@ -0,0 +1,533 @@
+		if err != nil {
+			merr.Append(fmt.Errorf("could not get user %s: %w", userID, err))
+			continue
+		}
+
+		sync, syncImage, err2 := scs.shouldUserSync(user, sd.task.channelID, sd.rc)
+		if err2 != nil {
+			merr.Append(fmt.Errorf("could not check should sync user %s: %w", userID, err))
+			continue
+		}
+
+		if sync {
+			sd.users[user.Id] = sanitizeUserForSync(user)
+		}
+
+		if syncImage {
+			sd.profileImages[user.Id] = sanitizeUserForSync(user)
+		}
+
+		// if this was a mention then put the real username in place of the username+remotename, but only
+		// when sending to the remote that the user belongs to.
+		if v.post != nil && user.RemoteId != nil && *user.RemoteId == sd.rc.RemoteId {
+			fixMention(v.post, v.mentionMap, user)
+		}
+	}
+	return merr.ErrorOrNil()
+}
+
+// fetchPostAttachmentsForSync populates the sync data with any file attachments for new posts.
+func (scs *Service) fetchPostAttachmentsForSync(sd *syncData) error {
+	merr := merror.New()
+	for _, post := range sd.posts {
+		fis, err := scs.server.GetStore().FileInfo().GetForPost(post.Id, false, true, true)
+		if err != nil {
+			merr.Append(fmt.Errorf("could not get file attachment info for post %s: %w", post.Id, err))
+			continue
+		}
+
+		for _, fi := range fis {
+			if scs.shouldSyncAttachment(fi, sd.rc) {
+				sd.attachments = append(sd.attachments, attachment{fi: fi, post: post})
+			}
+		}
+	}
+	return merr.ErrorOrNil()
+}
+
+// filterPostsforSync removes any posts that do not need to sync.
+func (scs *Service) filterPostsForSync(sd *syncData) {
+	filtered := make([]*model.Post, 0, len(sd.posts))
+
+	for _, p := range sd.posts {
+		// Don't resend an existing post where only the reactions changed.
+		// Posts we must send:
+		//   - new posts (EditAt == 0)
+		//   - edited posts (EditAt >= LastPostUpdateAt)
+		//   - deleted posts (DeleteAt > 0)
+		if p.EditAt > 0 && p.EditAt < sd.scr.LastPostUpdateAt && p.DeleteAt == 0 {
+			continue
+		}
+
+		// Don't send a deleted post if it is just the original copy from an edit.
+		if p.DeleteAt > 0 && p.OriginalId != "" {
+			continue
+		}
+
+		// don't sync a post back to the remote it came from.
+		if p.GetRemoteID() == sd.rc.RemoteId {
+			continue
+		}
+
+		// parse out all permalinks in the message.
+		p.Message = scs.processPermalinkToRemote(p)
+
+		filtered = append(filtered, p)
+	}
+	sd.posts = filtered
+}
+
+// sendSyncData sends all the collected users, posts, reactions, images, and attachments to the
+// remote cluster.
+// The order of items sent is important: users -> attachments -> posts -> reactions -> profile images
+func (scs *Service) sendSyncData(sd *syncData) error {
+	merr := merror.New()
+
+	// send users
+	if len(sd.users) != 0 {
+		if err := scs.sendUserSyncData(sd); err != nil {
+			merr.Append(fmt.Errorf("cannot send user sync data: %w", err))
+		}
+	}
+
+	// send attachments
+	if len(sd.attachments) != 0 {
+		scs.sendAttachmentSyncData(sd)
```

Comment:
If an error is returned from the calling function, it causes the whole sync task to be retried (up to X times).  A decision was made to treat some failures as log-only, so that the rest of the sync can proceed. I agree this can be re-visited; the `SharedChannelAttachments` table can be used to mark a file attachment as dirty so it can be retried independently of the posts.

Reply:
Understood, totally makes sense if it was explicitly decided. Thanks for the explanation!! :)